ocpi-toolkit-2.2.1/src/main/kotlin/com/izivia/ocpi/toolkit/modules/credentials/CredentialsServer.kt

```diff
@@ -3,14 +3,27 @@ package com.izivia.ocpi.toolkit.modules.credentials
 import com.izivia.ocpi.toolkit.common.*
 import com.izivia.ocpi.toolkit.modules.credentials.domain.Credentials
 import com.izivia.ocpi.toolkit.modules.credentials.services.CredentialsServerService
+import com.izivia.ocpi.toolkit.modules.versions.domain.InterfaceRole
+import com.izivia.ocpi.toolkit.modules.versions.domain.ModuleID
+import com.izivia.ocpi.toolkit.modules.versions.domain.VersionNumber
+import com.izivia.ocpi.toolkit.modules.versions.repositories.MutableVersionsRepository
 import com.izivia.ocpi.toolkit.transport.TransportServer
 import com.izivia.ocpi.toolkit.transport.domain.HttpMethod
 
 class CredentialsServer(
     private val service: CredentialsServerService,
-    basePath: String = "/2.2.1/credentials"
-) : OcpiModuleServer(basePath) {
-    override suspend fun registerOn(transportServer: TransportServer) {
+    versionsRepository: MutableVersionsRepository,
```

Comment:
`versionsRepository: MutableVersionsRepository` should be nullable and be null by default everywhere. So if one wants to opt in automatic endpoint registration they can, but it is not a must have.

In our case we will not use it because we have a list of available endpoints for each tenant

Reply:
just to clarify, and mostly out of curiosity, why do you need various endpoints? Is it just so that it is not obvious to connecting parties that is a multi tenant system?

You could still benefit from this approach, by moving responsibility of adjusting endpoint URL's to the VersionServer. That's partly what I tried to achieve here too, move all the HTTP specific stuff as close as possible to the HTTP components, as a repository shouldn't need to know such specifics as our external facing URL (ideally that's just some config you pass in as an env var). In your case, where you pretend to be at a certain URL for certain incoming connections, you could store the external baseUrl within your partner repo, provide a placeholder value in the baseUrl, and then have the VersionsServer replace(placeholder, actualValue).

Once #62 is implemented, you could do it in the VersionsService

Reply:
Each tenant has its sets of visible endpoints. We hold the tenant information in coroutines context. Moreover, we have separate server for registration and the rest. So the baseUrl of version / credentials is different from locations for instance.

Yes we could benefit from that, but we liked the fact that the way we implemented things was straightforward. We have a configuration that specifies with tenant can see what, and this config is only used at one place, in the VersionsRepository (all this logic is centralized in one place). We may change our minds later on, but as of now it works well that way

Reply:
In addition to that, we do not have access of versions repository from our other server, so we can not use automatic endpoint registration. Yes we could call endpoints from our other server to our credentials server, but if it's only to perform automatic endpoint registration, it is not worth the cost of development.

But we understand that this feature is really relevant for a lot of other cases ;)